mantle/kola/harness.go

```diff
@@ -1075,8 +1084,80 @@ func collectLogsExternalTest(h *harness.H, t *register.Test, tcluster cluster.Te
 	}
 }
 
+// Verifies that conflicts specified by each non-exclusive test are consistent
+// i.e test1 conflicts with test2 iff test2 conflicts with test1
+func conflictsSanityCheck(tests []*register.Test) error {
+	conflictMappings := make(map[string]map[string]bool)
+	for _, test := range tests {
+		for _, testConflict := range test.Conflicts {
+			if records, found := conflictMappings[test.Name]; found {
+				records[testConflict] = true
+			} else {
+				conflictMappings[test.Name] = map[string]bool{testConflict: true}
+			}
+		}
+	}
+
+	// Verify that if a test1 indicates test2 is a conflicting
+	// test, test2 also has done the same
+	for testName, records := range conflictMappings {
+		for conflictTest := range records {
+			if conflictRecords, found := conflictMappings[conflictTest]; found {
+				if _, found = conflictRecords[testName]; !found {
+					// conflictTest did not indicate testName as a conflict, but testName did
+					return fmt.Errorf("inconsistency: %v has indicated %v is conflicting but %v has not", testName, conflictTest, conflictTest)
+				}
+
+			} else {
+				// conflictTest has not indicated any conflicts
+				return fmt.Errorf("inconsistency: %v has indicated %v is conflicting but %v has not", testName, conflictTest, conflictTest)
+			}
+		}
+	}
+
+	return nil
+}
+
+func resolveTestConflicts(tests []*register.Test) [][]*register.Test {
+	// General idea:
+	//   - assign tests to the first available test bucket
+	//   - if a test bucket has a conflicting test, that bucket is unavailable
+	//   - if no buckets are available, create a new one
+	// [Greedy graph coloring] guarantees VMs used <= 1 + maximum number of conflicts one test has
+	assignments := make(map[string]int)
+	assignments[tests[0].Name] = 0
+	var buckets [][]*register.Test
+	buckets = append(buckets, []*register.Test{tests[0]})
+
+	for _, test := range tests[1:] {
+		// Find which of the test buckets are in use by conflicting tests
+		bucketInConflict := make(map[int]bool)
+		for _, testName := range test.Conflicts {
+			if bucket, found := assignments[testName]; found {
+				bucketInConflict[bucket] = true
+			}
+		}
+		// Find a test bucket that is not being used by a conflicting test
+		i := 0
+		for ; i < len(buckets); i++ {
+			if _, inConflict := bucketInConflict[i]; !inConflict {
+				assignments[test.Name] = 0
+				buckets[i] = append(buckets[i], test)
+				break
+			}
+		}
+		// Add a new bucket for this test, since all existing ones are being used
+		// by conflicting tests
+		if i == len(buckets) {
+			buckets = append(buckets, []*register.Test{test})
+		}
+
+	}
+	return buckets
+}
+
 // Create a parent test that runs non-exclusive tests as subtests
-func makeNonExclusiveTest(tests []*register.Test, flight platform.Flight) register.Test {
+func makeNonExclusiveTest(bucketNum int, tests []*register.Test, flight platform.Flight) register.Test {
```

Comment:
```suggestion
func makeNonExclusiveTest(bucket int, tests []*register.Test, flight platform.Flight) register.Test {
```

seems cleaner to just call this `bucket`.